javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure

```
[ERROR] .... Could not pass event PlayerQuitEvent to DiscordSRV v1.21.2
github.scarsz.discordsrv.dependencies.kevinsawicki.http.HttpRequest$HttpRequestException: javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
    at github.scarsz.discordsrv.dependencies.kevinsawicki.http.HttpRequest.code(HttpRequest.java:1562) ~[?:?]
    at github.scarsz.discordsrv.dependencies.mojang.Mojang.fetch(Mojang.java:81) ~[?:?]
    at github.scarsz.discordsrv.DiscordSRV.getAvatarUrl(DiscordSRV.java:1892) ~[?:?]
    at github.scarsz.discordsrv.DiscordSRV.getAvatarUrl(DiscordSRV.java:1883) ~[?:?]
    at github.scarsz.discordsrv.DiscordSRV.sendLeaveMessage(DiscordSRV.java:1697) ~[?:?]
    at github.scarsz.discordsrv.listeners.PlayerJoinLeaveListener.PlayerQuitEvent(PlayerJoinLeaveListener.java:132) ~[?:?]
    at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor85.execute(Unknown Source) ~[?:?]
    at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69) ~[patched_1.16.4.jar:git-Paper-406]
    at org.bukkit.plugin.EventExecutor$$Lambda$3676.0000000014944BD0.execute(Unknown Source) ~[?:?]
    at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80) ~[patched_1.16.4.jar:git-Paper-406]
    at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[patched_1.16.4.jar:git-Paper-406]
    at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:607) ~[patched_1.16.4.jar:git-Paper-406]
    at net.minecraft.server.v1_16_R3.PlayerList.disconnect(PlayerList.java:513) ~[patched_1.16.4.jar:git-Paper-406]
    at net.minecraft.server.v1_16_R3.PlayerConnection.a(PlayerConnection.java:1690) ~[patched_1.16.4.jar:git-Paper-406]
    at net.minecraft.server.v1_16_R3.NetworkManager.handleDisconnection(NetworkManager.java:502) ~[patched_1.16.4.jar:git-Paper-406]
    at net.minecraft.server.v1_16_R3.ServerConnection.c(ServerConnection.java:165) ~[patched_1.16.4.jar:git-Paper-406]
    at net.minecraft.server.v1_16_R3.MinecraftServer.b(MinecraftServer.java:1376) ~[patched_1.16.4.jar:git-Paper-406]
    at net.minecraft.server.v1_16_R3.DedicatedServer.b(DedicatedServer.java:377) ~[patched_1.16.4.jar:git-Paper-406]
    at net.minecraft.server.v1_16_R3.MinecraftServer.a(MinecraftServer.java:1204) ~[patched_1.16.4.jar:git-Paper-406]
    at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:992) ~[patched_1.16.4.jar:git-Paper-406]
    at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$a$0(MinecraftServer.java:173) ~[patched_1.16.4.jar:git-Paper-406]
    at net.minecraft.server.v1_16_R3.MinecraftServer$$Lambda$3147.0000000028186700.run(Unknown Source) ~[?:?]
    at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
    at sun.security.ssl.Alert.createSSLException(Unknown Source) ~[?:?]
    at sun.security.ssl.Alert.createSSLException(Unknown Source) ~[?:?]
    at sun.security.ssl.TransportContext.fatal(Unknown Source) ~[?:?]
    at sun.security.ssl.Alert$AlertConsumer.consume(Unknown Source) ~[?:?]
    at sun.security.ssl.TransportContext.dispatch(Unknown Source) ~[?:?]
    at sun.security.ssl.SSLTransport.decode(Unknown Source) ~[?:?]
    at sun.security.ssl.SSLSocketImpl.decode(Unknown Source) ~[?:?]
    at sun.security.ssl.SSLSocketImpl.readHandshakeRecord(Unknown Source) ~[?:?]
    at sun.security.ssl.SSLSocketImpl.startHandshake(Unknown Source) ~[?:?]
    at sun.net.www.protocol.https.HttpsClient.afterConnect(Unknown Source) ~[?:?]
    at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(Unknown Source) ~[?:?]
    at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source) ~[?:?]
    at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source) ~[?:?]
    at java.net.HttpURLConnection.getResponseCode(Unknown Source) ~[?:?]
    at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(Unknown Source) ~[?:?]
    at github.scarsz.discordsrv.dependencies.kevinsawicki.http.HttpRequest.code(HttpRequest.java:1560) ~[?:?]
... 22 more
```

--- a/pom.xml
+++ b/pom.xml
@@ -506,13 +506,6 @@
             <version>5.2.7.RELEASE</version>
         </dependency>
 
-        <!-- mojang api for avatar textures -->
-        <dependency>
-            <groupId>github.scarsz</groupId>
-            <artifactId>mojang</artifactId>
-            <version>1.0</version>
-        </dependency>
-
         <!-- text parsing & serialization -->
 
         <!-- Discord <-> Minecraft reserializer -->

--- a/src/main/java/github/scarsz/discordsrv/DiscordSRV.java
+++ b/src/main/java/github/scarsz/discordsrv/DiscordSRV.java
@@ -1899,7 +1899,7 @@ public class DiscordSRV extends JavaPlugin {
     private static String constructAvatarUrl(String username, UUID uuid, String texture) {
         boolean offline = uuid == null || PlayerUtil.uuidIsOffline(uuid);
         OfflinePlayer player = null;
-        if (username != null && offline) {
+        if (StringUtils.isNotBlank(username) && offline) {
             // resolve username to player/uuid
             //TODO resolve name to online uuid when offline player is present
             // (can't do it by calling Bukkit.getOfflinePlayer(username).getUniqueId() because bukkit just returns the offline-mode CraftPlayer)
@@ -1907,7 +1907,7 @@ public class DiscordSRV extends JavaPlugin {
             uuid = player.getUniqueId();
             offline = PlayerUtil.uuidIsOffline(uuid);
         }
-        if (username == null && uuid != null) {
+        if (StringUtils.isBlank(username) && uuid != null) {
             // resolve uuid to player/username
             player = Bukkit.getOfflinePlayer(uuid);
             username = player.getName();

--- a/src/main/java/github/scarsz/discordsrv/DiscordSRV.java
+++ b/src/main/java/github/scarsz/discordsrv/DiscordSRV.java
@@ -148,6 +148,7 @@ public class DiscordSRV extends JavaPlugin {
     public static boolean updateIsAvailable = false;
     public static boolean updateChecked = false;
     public static boolean invalidBotToken = false;
+    private static boolean offlineUuidAvatarUrlNagged = false;
     public static String version = "";
 
     // Managers
@@ -1896,25 +1897,41 @@ public class DiscordSRV extends JavaPlugin {
         return avatarUrl;
     }
     private static String constructAvatarUrl(String username, UUID uuid, String texture) {
+        boolean offline = uuid == null || PlayerUtil.uuidIsOffline(uuid);
         OfflinePlayer player = null;
-        if (username != null && uuid == null) {
+        if (username != null && offline) {
+            // resolve username to player/uuid
+            //TODO resolve name to online uuid when offline player is present
+            // (can't do it by calling Bukkit.getOfflinePlayer(username).getUniqueId() because bukkit just returns the offline-mode CraftPlayer)
             player = Bukkit.getOfflinePlayer(username);
             uuid = player.getUniqueId();
+            offline = PlayerUtil.uuidIsOffline(uuid);
         }
         if (username == null && uuid != null) {
+            // resolve uuid to player/username
             player = Bukkit.getOfflinePlayer(uuid);
             username = player.getName();
         }
         if (StringUtils.isBlank(texture) && player != null && player.isOnline()) {
+            // grab texture placeholder from player if online
             texture = NMSUtil.getTexture(player.getPlayer());
         }
 
         String avatarUrl = DiscordSRV.config().getString("AvatarUrl");
-        if (StringUtils.isBlank(avatarUrl))
-            avatarUrl = "https://crafatar.com/avatars/{uuid-nodashes}.png?size={size}&overlay#{texture}";
+        if (StringUtils.isBlank(avatarUrl)) {
+            avatarUrl = !offline ? "https://crafatar.com/avatars/{uuid-nodashes}.png?size={size}&overlay#{texture}"
+                                 : "https://minotar.net/helm/{username}/{size}.png#{texture}";
+        }
+        if (offline && !avatarUrl.contains("{username}")) {
+            if (!offlineUuidAvatarUrlNagged) {
+                DiscordSRV.error("Your AvatarUrl does not contain the {username} placeholder even though this server is using offline UUIDs.");
+                DiscordSRV.error("You should set your AvatarUrl to https://minotar.net/helm/{username}/{size}.png#{texture} to get avatars to work.");
+                offlineUuidAvatarUrlNagged = true;
+            }
+        }
 
         avatarUrl = avatarUrl
-                .replace("{texture}", texture)
+                .replace("{texture}", texture != null ? texture : "")
                 .replace("{username}", username)
                 .replace("{uuid}", uuid != null ? uuid.toString() : "")
                 .replace("{uuid-nodashes}", uuid.toString().replace("-", ""))

--- a/src/main/java/github/scarsz/discordsrv/DiscordSRV.java
+++ b/src/main/java/github/scarsz/discordsrv/DiscordSRV.java
@@ -58,8 +58,6 @@ import github.scarsz.discordsrv.objects.managers.link.JdbcAccountLinkManager;
 import github.scarsz.discordsrv.objects.metrics.BStats;
 import github.scarsz.discordsrv.objects.threads.*;
 import github.scarsz.discordsrv.util.*;
-import github.scarsz.mojang.Mojang;
-import github.scarsz.mojang.exception.ProfileFetchException;
 import lombok.Getter;
 import net.dv8tion.jda.api.*;
 import net.dv8tion.jda.api.entities.*;
@@ -116,7 +114,7 @@ import org.minidns.record.Record;
 import javax.annotation.CheckReturnValue;
 import javax.net.ssl.SSLContext;
 import javax.security.auth.login.LoginException;
-import java.awt.Color;
+import java.awt.*;
 import java.io.*;
 import java.lang.reflect.Field;
 import java.lang.reflect.InvocationTargetException;
@@ -124,6 +122,7 @@ import java.lang.reflect.Method;
 import java.net.*;
 import java.nio.charset.StandardCharsets;
 import java.sql.SQLException;
+import java.util.List;
 import java.util.*;
 import java.util.concurrent.*;
 import java.util.function.BiFunction;
@@ -586,8 +585,6 @@ public class DiscordSRV extends JavaPlugin {
         } catch (IOException e) {
             error(e);
         }
-
-        Mojang.setUserAgent("DiscordSRV/" + getDescription().getVersion());
 
         requireLinkModule = new RequireLinkModule();
 
@@ -1879,53 +1876,50 @@ public class DiscordSRV extends JavaPlugin {
         return messageBuilder.isEmpty() ? null : messageBuilder.build();
     }
 
-    public static String getAvatarUrl(OfflinePlayer player) {
-        return getAvatarUrl(player.getName());
+    public static String getAvatarUrl(String username, UUID uuid) {
+        String avatarUrl = constructAvatarUrl(username, uuid, "");
+        avatarUrl = PlaceholderUtil.replacePlaceholders(avatarUrl);
+        return avatarUrl;
     }
-    public static String getAvatarUrl(String username) {
-        if (username.startsWith("*")) {
-            // geyser
-            username = username.substring(1);
-        }
-
-        try {
-            return getAvatarUrl(Mojang.fetch(username), username);
-        } catch (ProfileFetchException e) {
-            DiscordSRV.error("Failed to retrieve avatar for player " + username, e);
-            return null;
-        }
-    }
-    public static String getAvatarUrl(UUID uuid) {
-        String name = Bukkit.getOfflinePlayer(uuid).getName();
-        try {
-            return getAvatarUrl(Mojang.fetch(uuid), name);
-        } catch (ProfileFetchException e) {
-            DiscordSRV.error("Failed to retrieve avatar for " + (name != null ? "player " + name : "unknown player"), e);
-            return null;
-        }
-    }
-    public static String getAvatarUrl(Mojang.GameProfile profile, String username) {
-        String avatarUrl = DiscordSRV.config().getString("AvatarUrl");
-        if (StringUtils.isBlank(avatarUrl)) avatarUrl = "https://crafatar.com/avatars/{uuid-nodashes}?overlay&size={size}";
-        if (profile != null) {
-            avatarUrl = avatarUrl
-                    .replace("{texture}", profile.getSkin())
-                    .replace("{username}", profile.getName())
-                    .replace("{uuid}", profile.getUuid().toString())
-                    .replace("{uuid-nodashes}", profile.getUuid().toString().replace("-", ""))
-                    .replace("{size}", "128");
-            avatarUrl = PlaceholderUtil.replacePlaceholdersToDiscord(avatarUrl, Bukkit.getOfflinePlayer(profile.getUuid()));
+    private static String getAvatarUrl(OfflinePlayer player) {
+        if (player.isOnline()) {
+            return getAvatarUrl(player.getPlayer());
         } else {
-            avatarUrl = avatarUrl
-                    .replace("{texture}", "")
-                    .replace("{username}", username)
-                    .replace("{uuid}", "c06f8906-4c8a-4911-9c29-ea1dbd1aab82")
-                    .replace("{uuid-nodashes}", "c06f8906-4c8a-4911-9c29-ea1dbd1aab82".replace("-", ""))
-                    .replace("{size}", "128");
-            avatarUrl = PlaceholderUtil.replacePlaceholdersToDiscord(avatarUrl);
+            String avatarUrl = constructAvatarUrl(player.getName(), player.getUniqueId(), "");
+            avatarUrl = PlaceholderUtil.replacePlaceholdersToDiscord(avatarUrl, player);
+            return avatarUrl;
+        }
+    }
+    public static String getAvatarUrl(Player player) {
+        String avatarUrl = constructAvatarUrl(player.getName(), player.getUniqueId(), NMSUtil.getTexture(player));
+        avatarUrl = PlaceholderUtil.replacePlaceholdersToDiscord(avatarUrl, player);
+        return avatarUrl;
+    }
+    private static String constructAvatarUrl(String username, UUID uuid, String texture) {
+        OfflinePlayer player = null;
+        if (username != null && uuid == null) {
+            player = Bukkit.getOfflinePlayer(username);
+            uuid = player.getUniqueId();
+        }
+        if (username == null && uuid != null) {
+            player = Bukkit.getOfflinePlayer(uuid);
+            username = player.getName();
+        }
+        if (StringUtils.isBlank(texture) && player != null && player.isOnline()) {
+            texture = NMSUtil.getTexture(player.getPlayer());
         }
 
-        debug("Avatar URL generated: " + avatarUrl);
+        String avatarUrl = DiscordSRV.config().getString("AvatarUrl");
+        if (StringUtils.isBlank(avatarUrl))
+            avatarUrl = "https://crafatar.com/avatars/{uuid-nodashes}.png?size={size}&overlay#{texture}";
+
+        avatarUrl = avatarUrl
+                .replace("{texture}", texture)
+                .replace("{username}", username)
+                .replace("{uuid}", uuid != null ? uuid.toString() : "")
+                .replace("{uuid-nodashes}", uuid.toString().replace("-", ""))
+                .replace("{size}", "128");
+
         return avatarUrl;
     }
 

--- a/src/main/java/github/scarsz/discordsrv/DiscordSRV.java
+++ b/src/main/java/github/scarsz/discordsrv/DiscordSRV.java
@@ -1930,6 +1930,11 @@ public class DiscordSRV extends JavaPlugin {
             }
         }
 
+        if (username.startsWith("*")) {
+            // geyser adds * to beginning of it's usernames
+            username = username.substring(1);
+        }
+
         avatarUrl = avatarUrl
                 .replace("{texture}", texture != null ? texture : "")
                 .replace("{username}", username)

--- a/src/main/java/github/scarsz/discordsrv/hooks/chat/VentureChatHook.java
+++ b/src/main/java/github/scarsz/discordsrv/hooks/chat/VentureChatHook.java
@@ -172,7 +172,7 @@ public class VentureChatHook implements ChatHook {
             webhookUsername = PlaceholderUtil.replacePlaceholders(webhookUsername);
             webhookUsername = DiscordUtil.strip(webhookUsername);
 
-            WebhookUtil.deliverMessage(destinationChannel, webhookUsername, DiscordSRV.getAvatarUrl(username), message, null);
+            WebhookUtil.deliverMessage(destinationChannel, webhookUsername, DiscordSRV.getAvatarUrl(username, chatPlayer != null ? chatPlayer.getUUID() : null), message, null);
         }
     }
 

--- a/src/main/java/github/scarsz/discordsrv/util/NMSUtil.java
+++ b/src/main/java/github/scarsz/discordsrv/util/NMSUtil.java
@@ -1,0 +1,151 @@
+/*-
+ * LICENSE
+ * DiscordSRV
+ * -------------
+ * Copyright (C) 2016 - 2021 Austin "Scarsz" Shapiro
+ * -------------
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * END
+ */
+
+package github.scarsz.discordsrv.util;
+
+import org.apache.commons.codec.binary.Base64;
+import org.bukkit.Bukkit;
+import org.bukkit.entity.Player;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class NMSUtil {
+
+    private static final Pattern TEXTURE_URL_PATTERN = Pattern.compile("https?:\\/\\/.+(?<texture>\\w{64})\\\",");
+
+    protected static String versionPrefix = "";
+
+    protected static Class<?> class_CraftPlayer;
+    protected static Class<?> class_GameProfile;
+    protected static Class<?> class_GameProfileProperty;
+    protected static Class<?> class_EntityPlayer;
+    protected static Method method_CraftPlayer_getHandle;
+    protected static Method method_EntityPlayer_getGameProfile;
+    protected static Method method_GameProfile_getProperties;
+    protected static Field field_PropertyMap_properties;
+    protected static Field field_GameProfileProperty_value;
+
+    static {
+        String className = Bukkit.getServer().getClass().getName();
+        String[] packages = className.split("\\.");
+        if (packages.length == 5) {
+            versionPrefix = packages[3] + ".";
+        }
+
+        try {
+            class_EntityPlayer = fixBukkitClass("net.minecraft.server.EntityPlayer");
+            method_EntityPlayer_getGameProfile = class_EntityPlayer.getMethod("getProfile");
+
+            class_CraftPlayer = fixBukkitClass("org.bukkit.craftbukkit.entity.CraftPlayer");
+            method_CraftPlayer_getHandle = class_CraftPlayer.getMethod("getHandle");
+
+            class_GameProfile = getClass("com.mojang.authlib.GameProfile");
+            class_GameProfileProperty = getClass("com.mojang.authlib.properties.Property");
+            if (class_GameProfile == null) {
+                class_GameProfile = getClass("net.minecraft.util.com.mojang.authlib.GameProfile");
+                class_GameProfileProperty = getClass("net.minecraft.util.com.mojang.authlib.properties.Property");
+            }
+            method_GameProfile_getProperties = class_GameProfile.getMethod("getProperties");
+            field_GameProfileProperty_value = class_GameProfileProperty.getDeclaredField("value");
+            field_GameProfileProperty_value.setAccessible(true);
+            field_PropertyMap_properties = method_GameProfile_getProperties.getReturnType().getDeclaredField("properties");
+            field_PropertyMap_properties.setAccessible(true);
+        } catch (Throwable e) {
+            e.printStackTrace();
+        }
+    }
+
+    public static Class<?> getClass(String className) {
+        Class<?> result = null;
+        try {
+            result = NMSUtil.class.getClassLoader().loadClass(className);
+        } catch (Exception ignored) {}
+        return result;
+    }
+
+    public static Class<?> fixBukkitClass(String className) throws ClassNotFoundException {
+        if (!versionPrefix.isEmpty()) {
+            className = className.replace("org.bukkit.craftbukkit.", "org.bukkit.craftbukkit." + versionPrefix);
+            className = className.replace("net.minecraft.server.", "net.minecraft.server." + versionPrefix);
+        }
+
+        return NMSUtil.class.getClassLoader().loadClass(className);
+    }
+
+    public static Object getHandle(Player player) {
+        Object handle = null;
+        try {
+            handle = method_CraftPlayer_getHandle.invoke(player);
+        } catch (Throwable e) {
+            e.printStackTrace();
+        }
+        return handle;
+    }
+
+    public static Object getGameProfile(Player player) {
+        Object profile = null;
+        Object handle = getHandle(player);
+        if (handle != null) {
+            try {
+                profile = method_EntityPlayer_getGameProfile.invoke(handle);
+            } catch (Throwable e) {
+                e.printStackTrace();
+            }
+        }
+        return profile;
+    }
+
+    public static Object getTextureProperty(Object propertyMap) {
+        Object property = null;
+        try {
+            Object multi = NMSUtil.field_PropertyMap_properties.get(propertyMap);
+            //noinspection rawtypes
+            property = ((Iterable) multi.getClass()
+                        .getMethod("get", Object.class)
+                        .invoke(multi, "textures"))
+                    .iterator().next();
+        } catch (Throwable e) {
+            e.printStackTrace();
+        }
+        return property;
+    }
+
+    public static String getTexture(Player player) {
+        String texture = null;
+        try {
+            Object profile = getGameProfile(player);
+            Object propertyMap = method_GameProfile_getProperties.invoke(profile);
+            Object textureProperty = getTextureProperty(propertyMap);
+            String textureB64 = (String) field_GameProfileProperty_value.get(textureProperty);
+            String textureData = new String(Base64.decodeBase64(textureB64));
+            Matcher matcher = TEXTURE_URL_PATTERN.matcher(textureData);
+            if (matcher.find()) texture = matcher.group("texture");
+        } catch (Throwable e) {
+            e.printStackTrace();
+        }
+        return texture;
+    }
+
+}

--- a/src/main/java/github/scarsz/discordsrv/util/NMSUtil.java
+++ b/src/main/java/github/scarsz/discordsrv/util/NMSUtil.java
@@ -28,6 +28,7 @@ import org.bukkit.entity.Player;
 
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
+import java.util.Iterator;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
@@ -95,57 +96,57 @@ public class NMSUtil {
     }
 
     public static Object getHandle(Player player) {
-        Object handle = null;
         try {
-            handle = method_CraftPlayer_getHandle.invoke(player);
+            return method_CraftPlayer_getHandle.invoke(player);
         } catch (Throwable e) {
             e.printStackTrace();
         }
-        return handle;
+        return null;
     }
 
     public static Object getGameProfile(Player player) {
-        Object profile = null;
         Object handle = getHandle(player);
         if (handle != null) {
             try {
-                profile = method_EntityPlayer_getGameProfile.invoke(handle);
+                return method_EntityPlayer_getGameProfile.invoke(handle);
             } catch (Throwable e) {
                 e.printStackTrace();
             }
         }
-        return profile;
+        return null;
     }
 
     public static Object getTextureProperty(Object propertyMap) {
-        Object property = null;
         try {
             Object multi = NMSUtil.field_PropertyMap_properties.get(propertyMap);
             //noinspection rawtypes
-            property = ((Iterable) multi.getClass()
-                        .getMethod("get", Object.class)
-                        .invoke(multi, "textures"))
-                    .iterator().next();
+            Iterator it = ((Iterable) multi.getClass()
+                    .getMethod("get", Object.class)
+                    .invoke(multi, "textures")).iterator();
+            if (it.hasNext()) {
+                return it.next();
+            }
         } catch (Throwable e) {
             e.printStackTrace();
         }
-        return property;
+        return null;
     }
 
     public static String getTexture(Player player) {
-        String texture = null;
         try {
             Object profile = getGameProfile(player);
             Object propertyMap = method_GameProfile_getProperties.invoke(profile);
             Object textureProperty = getTextureProperty(propertyMap);
-            String textureB64 = (String) field_GameProfileProperty_value.get(textureProperty);
-            String textureData = new String(Base64.decodeBase64(textureB64));
-            Matcher matcher = TEXTURE_URL_PATTERN.matcher(textureData);
-            if (matcher.find()) texture = matcher.group("texture");
+            if (textureProperty != null) {
+                String textureB64 = (String) field_GameProfileProperty_value.get(textureProperty);
+                String textureData = new String(Base64.decodeBase64(textureB64));
+                Matcher matcher = TEXTURE_URL_PATTERN.matcher(textureData);
+                if (matcher.find()) return matcher.group("texture");
+            }
         } catch (Throwable e) {
             e.printStackTrace();
         }
-        return texture;
+        return null;
     }
 
 }

--- a/src/main/java/github/scarsz/discordsrv/util/PlayerUtil.java
+++ b/src/main/java/github/scarsz/discordsrv/util/PlayerUtil.java
@@ -218,4 +218,13 @@ public class PlayerUtil {
         return Character.isWhitespace(character) || character == '@';
     }
 
+    /**
+     * Returns whether the passed UUID is a v3 UUID. Offline UUIDs are v3, online are v4.
+     * @param uuid the UUID to check
+     * @return whether the UUID is a v3 UUID & thus is offline
+     */
+    public static boolean uuidIsOffline(UUID uuid) {
+        return uuid.toString().charAt(14) == '3';
+    }
+
 }

--- a/src/main/java/github/scarsz/discordsrv/util/PlayerUtil.java
+++ b/src/main/java/github/scarsz/discordsrv/util/PlayerUtil.java
@@ -224,7 +224,7 @@ public class PlayerUtil {
      * @return whether the UUID is a v3 UUID & thus is offline
      */
     public static boolean uuidIsOffline(UUID uuid) {
-        return uuid.toString().charAt(14) == '3';
+        return uuid.version() == 3;
     }
 
 }

